Opt-out updated bootimage for GCP and AWS (#93065)
openshift/openshift-docs: machine_configuration/mco-update-boot-images.adoc

toc::[]

The Machine Config Operator (MCO) uses a boot image to scale up {op-system-first} nodes. Whether the Machine Config Operator (MCO) updates the boot image in your cluster depends upon the platform your cluster is built upon.

> * For {gcp-first}, starting {product-title} 4.16, the MCO updates the boot image with each cluster update by default. New nodes scale with the same version as the cluster. 

Author: Oh, of course. Thanks.

> * For {aws-first}, starting {product-title} 4.17, the MCO updates the boot image with each cluster update by default. New nodes scale with the same version as the cluster. 

Reviewer: same issue as https://github.com/openshift/openshift-docs/pull/93065/files#r2091601195

* For all other platforms, by default the MCO does not update the boot image with each cluster update. For example, if your cluster was originally created with {product-title} 4.12, the boot image that the MCO uses to create nodes is the same 4.12 version, even if your cluster is at a later version. If the cluster is later upgraded to 4.13 or later, new nodes continue to scale with the same 4.12 image.

For {gcp-short} and {aws-short}, you can disable the default behavior, if needed. When disabled, the boot image no longer updates with the cluster. 

[NOTE]
====
The ability to configure this behavior is available for only {gcp-short} and {aws-short} clusters. It is not supported for clusters managed by the {cluster-capi-operator}.
====

> You might want to disable the use of updated boot images, for example, if you are not using the default user data secret, named `worker-user-data`, in your machine set, or you have modified the `worker-user-data` secret. This is because the MCO updates the machine set to use a managed version of the secret. By using the updated boot images feature, you are giving up the capability to customize the secret stored in the machine set object. 

Reviewer: So this is an old behavior that is no longer done in 4.19.0. This was part of the [stub upgrade work](https://github.com/openshift/machine-config-operator/pull/4885) we discussed. In 4.18 or lower, we used to hardcode the machineset to `worker-user-data-managed`. But with the PR above, we will keep the existing `*-user-data` referenced in the machineset and upgrade it to spec 3 if required. If it is already spec 3, the secret will be untouched. 

We could probably drop this paragraph completely and deal with it in https://issues.redhat.com/browse/OSDOCS-14592.

Author: @djoshy Do we have a reason why one would opt out that was can put in its place?

Reviewer: Not sure I follow your question there, are you asking if there a different reason to opt-out, now that we no longer hardcode the secrets?

Author: Yes.

Reviewer: My hope is that https://issues.redhat.com/browse/OSDOCS-14592 would explain that an Ignition upgrade is happening on the the secret(instead of upgrading) and any remediation to be done if it fails. So I suppose we could add those docs here, and remove the current paragraph about hardcoding?

Reviewer: Gotcha, then I think +1 to dropping this paragraph about hardcoding the secret.

However, using an older boot image could cause the following issues:

* Extra time to start nodes
* Certificate expiration issues
* Version skew issues

> For information on how to disable the default behavior, see "Disabling updated boot images". If you disable the default behavior, you can re-enable the default behavior at any time. For more information, see "Re-enabling updated boot images".

Reviewer: > "Disabling updated boot images"
Is this meant to be a hyperlink?

Author: We aren't allowed to include links in our "module" files, like this one.

How the cluster behaves after disabling or re-enabling the default behavior, depends upon when you made the change, including the following scenarios:

* If you disable the behavior before updating to a new {product-title} version:

> ** The boot image version used by the machine sets remains at the same after the update. 

Reviewer: ```suggestion
** The boot image version used by the machine sets remains the same, matching the version you upgraded from. 
```
Not sure how to phrase this, but the I just want to make sure that "the boot images will match the value it was at when the feature was disabled"

> ** The boot image version on the existing nodes in your cluster also remains the same after the update. 

Reviewer: I'm not sure this line is necessary. Conceptually, a boot image only matters for a node that is yet to be born. We can't change the boot image of a node that already exists, as that node will never go through first boot again.

Author: @djoshy I wanted to make it clear to the user that the existing nodes don't change. I can remove it if you think it is too obvious.

Reviewer: Gotcha, I agree that we should be clear about it. Maybe stating that at the top might be better than doing a bullet for it in each case? I fear stating it separately in each case implies that there is a possibility it could be changed(whether by the MCO or some other way), and I just want to be very clear that conceptually, it doesn't make sense for an existing node.

Reviewer: These lines are still present, so it seems unresolved to me?

** Any new nodes use that same version.

* If you disable the behavior after updating to a new {product-title} version: 
** The boot image version used by the machine sets is converted to match the updated {product-title} version. 

> ** The boot image version on the existing nodes in your cluster remains unchanged. 

Reviewer: same concern as https://github.com/openshift/openshift-docs/pull/93065/files#r2091637012

** Any new nodes use the updated {product-title} version when a machine set is scaled. 
** If you update to a later {product-title} version, the boot image version in the machine sets remains at the current version and are not updated with the cluster.

* If you re-enable the behavior: 
** The boot image version used by the machine sets is converted to the current version, if different. 
** The boot image version on the nodes in your cluster match the {product-title} version that was present when you disabled the behavior. 

> ** The boot image version on existing nodes does not change. 

Reviewer: same issue as https://github.com/openshift/openshift-docs/pull/93065/files#r2091637012
